scripts/migrate-portals.js

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env node
+const fs = require('fs');
+const readline = require('readline');
+const { MatrixClient, LogService } = require('matrix-bot-sdk');
+
+LogService.setLevel('ERROR');
+
+const client = new MatrixClient(process.env.MATRIX_URL, process.env.MATRIX_AT);
+
+const dryRun = process.env.DRY_RUN ? process.env.DRY_RUN === 'true' : true;
+
+const rl = readline.createInterface({
+    input: fs.createReadStream(process.env.ROOM_STORE || './room-store.db'),
+    output: process.stdout,
+    terminal: false
+});
+
+let portalRooms = [];
+
+rl.on('line', (line) => {
+    const roomEntry = JSON.parse(line);
+    if (!roomEntry.matrix_id) {
+        return;
+    }
+    if (!roomEntry || !roomEntry.data || !roomEntry.data.portal) {
+        return;
+    }
+    const gitterRoomId = roomEntry.remote_id;
+    const oldMatrixRoomId = roomEntry.matrix_id;
+    const localpart = gitterRoomId.replace('/', '_');
+    const alias = `#${localpart}:gitter.im`;
+    portalRooms.push({gitterRoomId, oldMatrixRoomId, alias});
+});
+
+async function getNewRoomId(alias) {
+    try {
+        const targetRoomId = await client.resolveRoom(alias);
+        return { targetRoomId, joined: false };
+    } catch (ex) {
+        if (dryRun) {
+            return {
+                targetRoomId: '!noroomyet:gitter.im',
+                joined: false,
+            };
+        }
+        console.log("Joining new gitter room...");
+        await new Promise((r) => setTimeout(r, 5000)); // Joins cost more in rate limits
+        return {
+            targetRoomId: await client.joinRoom(alias),
```

Comment:
I'm confused why we want the bridge to join the new room alias 🤔 

Is this the only method of getting the internal ID?

Reply:
If I understood @turt2live 's [point](https://matrix.to/#/!vYjRCIoySolvxmhMJE:matrix.org/$gSEHtcQJrxwXpHZdOBGWL-ilUYRKn4AcedJoko8mUcc?via=matrix.org&via=vector.modular.im&via=half-shot.uk) correctly, clients (should) use the sender's server as a potential via target. We need to join the room we are pointing to in the upgrade to make matrix.org aware of it.

The other point is if the room doesn't yet exist on the bridge (nobody has spoken / nobody tried to join it), we need to execute a join in order for the new bridge to create the mapping.

Reply:
Sounds good.

In any case, just thought it was an extra step but doesn't hurt anything. We leave at the end anyway.

Reply:
it's unsafe to rely on the alias in this case by the way - there's no spec that says the alias *must* be moved, so you're stuck with the room ID in the tombstone and the server which sent that tombstone.

Reply:
>We need to join the room we are pointing to in the upgrade to make matrix.org aware of it.

This ends up being somewhat bunk as the server needs to remain joined to it for this to be useful, and we part after we join.